src/native.rs

```diff
@@ -1414,6 +1434,219 @@ impl super::Context for Context {
+    }
+
+    unsafe fn push_debug_group<S>(&self, source: u32, id: u32, message: S)
+    where
+        S: AsRef<str>,
+    {
+        let gl = &self.raw;
+        let msg = message.as_ref().as_bytes();
+        let length = msg.len() as i32;
+        gl.PushDebugGroup(source, id, length, msg.as_ptr() as *const i8);
+    }
+
+    unsafe fn pop_debug_group(&self) {
+        let gl = &self.raw;
+        gl.PopDebugGroup();
+    }
+
+    unsafe fn object_label<S>(&self, identifier: u32, name: u32, label: Option<S>)
+    where
+        S: AsRef<str>,
+    {
+        let gl = &self.raw;
+
+        match label {
+            Some(l) => {
+                let lbl = l.as_ref().as_bytes();
+                let length = lbl.len() as i32;
+                gl.ObjectLabel(identifier, name, length, lbl.as_ptr() as *const i8);
+            }
+            None => gl.ObjectLabel(identifier, name, 0, std::ptr::null()),
+        }
+    }
+
+    unsafe fn get_object_label(&self, identifier: u32, name: u32) -> String {
+        let gl = &self.raw;
+        let mut len = 0;
+        let mut label_buf = Vec::with_capacity(self.constants.max_label_length as usize);
+        gl.GetObjectLabel(
+            identifier,
+            name,
+            self.constants.max_label_length,
+            &mut len,
+            label_buf.as_mut_ptr(),
+        );
+        label_buf.set_len(len as usize);
+        std::ffi::CStr::from_ptr(label_buf.as_ptr())
+            .to_str()
+            .unwrap()
+            .to_owned()
+    }
+
+    unsafe fn object_ptr_label<S>(&self, sync: Self::Fence, label: Option<S>)
+    where
+        S: AsRef<str>,
+    {
+        let gl = &self.raw;
+
+        match label {
+            Some(l) => {
+                let lbl = l.as_ref().as_bytes();
+                let length = lbl.len() as i32;
+                gl.ObjectPtrLabel(
+                    sync as *mut std::ffi::c_void,
+                    length,
+                    lbl.as_ptr() as *const i8
+                );
+            }
+            None => gl.ObjectPtrLabel(sync as *mut std::ffi::c_void, 0, std::ptr::null()),
+        }
+    }
+
+    unsafe fn get_object_ptr_label(&self, sync: Self::Fence) -> String {
+        let gl = &self.raw;
+        let mut len = 0;
+        let mut label_buf = Vec::with_capacity(self.constants.max_label_length as usize);
+        gl.GetObjectPtrLabel(
+            sync as *mut std::ffi::c_void,
+            self.constants.max_label_length,
+            &mut len,
+            label_buf.as_mut_ptr(),
+        );
+        label_buf.set_len(len as usize);
+        std::ffi::CStr::from_ptr(label_buf.as_ptr())
+            .to_str()
+            .unwrap()
+            .to_owned()
+    }
+}
+
+extern "system" fn raw_debug_message_callback<F>(
+    source: u32,
+    gltype: u32,
+    id: u32,
+    severity: u32,
+    length: i32,
+    message: *const i8,
+    user_param: *mut std::ffi::c_void,
+)
+where
+    F: FnMut(u32, u32, u32, u32, &str)
+{
+    unsafe {
```

Comment:
I think we might need [`catch_unwind`](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html) around this block to prevent unwinding across the FFI boundary